backend/server.js

```diff
@@ -134,39 +179,82 @@ function addBackendRoutes (app, config, store) {
      grants. */
   app.post('/upload', function (req, res) {
     config.getUserConfig(req, function (err, userConfig) {
-      selectTarget (userConfig, req.body, function (err, target) {
-        if (err) return res.json({error: err.toString()});
+      selectTarget(userConfig, req.body, function (err, target) {
+        if (err) return res.json({ error: err.toString() });
         const s3client = upload.makeS3UploadClient(target);
-        const {s3Bucket, uploadPath: uploadDir} = target;
+        const { s3Bucket, uploadPath: uploadDir } = target;
+        const targetName = req.body.targetName;
         const id = Date.now().toString();
         const uploadPath = `${uploadDir}/${id}`;
         upload.getJsonUploadForm(s3client, s3Bucket, uploadPath, function (err, events) {
-          if (err) return res.json({error: err.toString()});
+          if (err) return res.json({ error: err.toString() });
           upload.getMp3UploadForm(s3client, s3Bucket, uploadPath, function (err, audio) {
-            if (err) return res.json({error: err.toString()});
+            if (err) return res.json({ error: err.toString() });
             const baseUrl = `https://${s3Bucket}.s3.amazonaws.com/${uploadPath}`;
             const player_url = `${config.playerUrl}?base=${encodeURIComponent(baseUrl)}`;
-            res.json({player_url, events, audio});
+            if (!config.enableOauth && config.database)
+              mysqlUtils.storeRecord(req.session.context.userId, targetName, baseUrl, id, config.mysqlConnPool);
+            res.json({ player_url, events, audio });
           });
         });
       });
     });
   });
 
+  app.post('/record/delete', checkLogin, function (req, res) {
+    const recordId = req.body.recordId;
+    if (req.session.context.isAdmin) {
+      deleteRecord();
+    } else {
+      mysqlUtils.userHavePrivileges(recordId, req.session.context.userId, config.database, function (err, isAllowed) {
+        if (!err && isAllowed) {
+          deleteRecord();
+        } else {
+          console.error("user is not allowed to delete the record");
+          res.redirect(req.get('Referrer'));
+        }
+      });
+    }
+
+    function deleteRecord() {
+      config.getUserConfig(req, function (err, userConfig) {
+        selectTarget(userConfig, userConfig['grants'][0], function (err, target) {
+          if (err) return res.json({ error: err.toString() });
+          const { s3Bucket, uploadPath: uploadDir } = target;
+          const s3Client = upload.makeS3Client(target);
+          const keys = [
+            { Key: `${uploadDir}/${recordId}.mp3` },
+            { Key: `${uploadDir}/${recordId}.json` },
+          ]
+          upload.deleteObject(s3Client, s3Bucket, keys).then(function (data) {
+            if (!config.enableOauth && config.database) {
+              mysqlUtils.deleteRecord(recordId, config.mysqlConnPool);
```

Comment:
`deleteRecord` is an async method. we are not giving a callback here. So I don't think we will catch the error (if any) ever. This will always result in returning success response even if the deletion was unsuccessful. Please confirm this

Reply:
done